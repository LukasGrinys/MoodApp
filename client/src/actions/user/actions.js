export const actions = {
    createUser : {
        request : 'CREATE_USER_REQUEST',
        receive : 'CREATE_USER_RECEIVE',
        reset: 'CREATE_USER_RESET'
    },
    loginUser : {
        request : 'LOGIN_USER_REQUEST',
        receive : 'LOGIN_USER_RECEIVE'
    },
    clearUserData : 'CLEAR_USER',
    userAuth: 'USER_AUTH',
    editUser : {
        request : 'EDIT_USER_REQUEST',
        receive : 'EDIT_USER_RECEIVE',
        reset : 'EDIT_USER_RESET'
    },
    changePassword : {
        request : 'CHANGE_PASSWORD_REQUEST',
        receive : 'CHANGE_PASSWORD_RECEIVE',
        reset : 'CHANGE_PASSWORD_RESET'
    }
}