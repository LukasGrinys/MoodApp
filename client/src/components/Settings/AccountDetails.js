import React from 'react';
import { Wrapper, Input, Label, Footer, ErrorMessage } from '../Forms/Forms';
import Button from '../Button';
import ErrorBox from '../ErrorBox';
import SuccessBox from '../SuccessBox';
import styles from './Settings.module.scss';
import { useEditAccount } from '../../hooks/Settings/useEditAccount';

const AccountDetails = ({
    email,
    firstName,
    lastName
}) => {
    const {
        isSubmitting,
        isEditingUser,
        editUserError,
        editUserSuccess,
        values,
        errors,
        touched,
        isValid,
        handleSubmit,
        controlEvents
    } = useEditAccount({
        firstName,
        lastName
    })

    return (
        <Wrapper>
            <div><strong>User: </strong> {email}</div><br/>
            <h3 className={styles.miniHeader}>Edit details:</h3>
            <Label>First name:</Label>
            <Input
                name="firstName"
                value={values.firstName}
                controlEvents={controlEvents}
            />
            {
                (errors.firstName && touched.firstName) && (
                    <ErrorMessage>
                        {errors.firstName}
                    </ErrorMessage>
                )
            }
            <Label>Last name:</Label>
            <Input
                name="lastName"
                value={values.lastName}
                controlEvents={controlEvents}    
            />
            {
                (errors.lastName && touched.lastName) && (
                    <ErrorMessage>
                        {errors.lastName}
                    </ErrorMessage>
                )
            }
            <Footer>
                <Button
                    disabled={isSubmitting || !isValid || isEditingUser}
                    handleClick={handleSubmit}
                >
                    Save changes
                </Button>
            </Footer>
            {
                !isEditingUser && editUserError && (
                    <ErrorBox>
                        {editUserError}
                    </ErrorBox>
                )
            }
            {
                !isEditingUser && editUserSuccess && (
                    <SuccessBox>
                        Account information was saved
                    </SuccessBox>
                )
            }
        </Wrapper>
    );
};

export default AccountDetails;