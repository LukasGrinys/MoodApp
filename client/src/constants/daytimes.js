export const daytimes = {
    morning: 'Morning',
    afternoon: 'Afternoon',
    evening: 'Evening',
    night: 'Night'
}