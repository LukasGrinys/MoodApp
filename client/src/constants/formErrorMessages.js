export const formErrorMessages = {
    required : 'Required field',
    email: 'Enter a proper e-mail address'
}