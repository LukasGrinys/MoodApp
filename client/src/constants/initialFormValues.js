export const initialFormValues = {
    email: '',
    password: '',
    firstName: '',
    lastName: '',
    confirmPassword: '',
    currentPassword: ''
}