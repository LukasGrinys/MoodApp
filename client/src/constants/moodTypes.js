const moodTypes = [
    {
        above: 7.5,
        below: 10,
        text: 'Great'
    },
    {
        above: 5,
        below: 7.5,
        text: 'Good'
    },
    {
        above: 3,
        below: 5,
        text: 'Average'
    },
    {
        above: 0,
        below: 3,
        text: 'Bad'
    }
];

export default moodTypes;