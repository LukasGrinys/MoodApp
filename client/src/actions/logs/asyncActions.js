import axios from 'axios';
import { actions } from './actions';

export const canUserLog = async (currentDate, currentTiming, user) => {
    const body = {
        userId: user,
        date: currentDate,
        timing: currentTiming
    };

    return async dispatch => {
        const { data } = await axios.post('/api/canLog', body);

        dispatch({type: actions.canLog.request})
        
        if (data && data.data) {
            dispatch({
                type: actions.canLog.receive,
                payload: data.data.canLog
            })

            return;
        }

        dispatch({
            type: actions.canLog.receive,
            payload: false
        })
    }
}

export const getLastLogs = async ({
    userId,
    limit = 9
}) => {
    const url = `/api/getLogs?id=${userId}&limit=${limit}`;

    return async dispatch => {
        dispatch({type: actions.getLastLogs.request});

        try {
            const { data } = await axios.get(url);
            const { success, error } = data;
    
            if (error) {
                dispatch({
                    type: actions.getLastLogs.receive,
                    payload: {
                        logs: [],
                        error: error,
                        isFetching: false
                    }
                });
    
                return;
            }
    
            if (success && data.data) {
                dispatch({
                    type: actions.getLastLogs.receive,
                    payload: {
                        logs: data.data,
                        error: null,
                        isFetching: false
                    }
                });
    
                return;
            }
        } catch (error) {
            dispatch({
                type: actions.getLastLogs.receive,
                payload: {
                    logs: [],
                    error: error,
                    isFetching: false
                }
            });
        }
    }
}

export const clearLastLogs = () => dispatch => dispatch({type: actions.getLastLogs.reset});

export const getLog = async logId => {
    return async dispatch => {
        const endpoint = `/api/getLog?id=${logId}`;

        dispatch({type: actions.getLog.request});

        try {
            const response = await axios.get(endpoint);
            
            const { error, data } = response.data;

            if (error) {
                dispatch({
                    type: actions.getLog.receive,
                    payload: {
                        isFetching : false,
                        error,
                        logData : {}
                    }
                });

                return;
            }
            
            dispatch({
                type: actions.getLog.receive,
                payload: {
                    isFetching : false,
                    error : false,
                    logData : data
                }
            });
        } catch (error) {
            dispatch({
                type: actions.getLog.receive,
                payload: {
                    isFetching : false,
                    error,
                    logData : {}
                }
            });
        }
    }
}

export const clearLog = () => dispatch => { dispatch({type: actions.getLog.reset}) }

export const getLogs = async ({userId, start, limit, currentLogs}) => {
    return async dispatch => {
        const endpoint = `/api/getLogs?id=${userId}${start ? `&skip=${start}` : ''}&limit=${limit}`;

        dispatch({ type: actions.getLogs.request });

        try {
            const response = await axios.get(endpoint);
            const { data, error } = response.data;

            if (error) {
                dispatch({
                    type: actions.getLogs.receive,
                    payload: {
                        isFetching: false,
                        error
                    }
                });

                return;
            }
            
            dispatch({
                type: actions.getLogs.receive,
                payload: {
                    isFetching: false,
                    error: false,
                    logs: currentLogs ? [...currentLogs, ...data] : data,
                    noLogsLeft: !data.length ? true : false
                }
            });
        } catch (error) {
            dispatch({
                type: actions.getLogs.receive,
                payload: {
                    isFetching: false,
                    error
                }
            });
        }
    }
}

export const clearLogs = () => dispatch => { dispatch({ type: actions.getLogs.reset }) }

export const postLog = async ({
    userId,
    date,
    timing,
    rating,
    text
}) => {
    const body = {
        userId,
        date,
        timing,
        rating,
        text
    }

    return async dispatch => {
        try {
            dispatch( {type: actions.postLog.request});
    
            const { data } = await axios.post('/api/postLog', body);
            const { error , success } = data;

            if (error || !success) {
                dispatch({
                    type: actions.postLog.receive,
                    payload: {
                        isPosting : false,
                        success : false,
                        error : error || 'Unknown error occured'
                    }
                });

                return;
            }

            dispatch({
                type: actions.postLog.receive,
                payload: {
                    isPosting : false,
                    success : true,
                    error : null
                }
            })
        } catch (error) {
            dispatch({
                type: actions.postLog.receive,
                payload: {
                    isPosting : false,
                    success : false,
                    error
                }
            })
        }
    }  
}

export const clearLogPost = () => dispatch => { dispatch( {type: actions.postLog.reset})};