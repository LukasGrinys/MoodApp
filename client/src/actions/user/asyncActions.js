import { actions } from './actions';
import axios from 'axios';
import { clearLogs, clearLastLogs } from '../logs/asyncActions';
import Cookie from 'universal-cookie';
const cookie = new Cookie();

export const createUser = async user => {
    return async dispatch => {
        dispatch({type: actions.createUser.request});

        try {
            const { data } = await axios.post('/api/users', user)
            const { success, error } = data;

            if (error || !success) {
                dispatch({
                    type: actions.createUser.receive,
                    payload: {
                        success: false,
                        error: error || 'Could not create user'
                    }
                });

                return;
            }
            
            if (success) {
                dispatch({
                    type: actions.createUser.receive,
                    payload: {
                        success: true,
                        error: null
                    }
                });
            }
        } catch (error) {
            dispatch({
                type: actions.createUser.receive,
                payload: {
                    success: false,
                    error: error || 'Could not create user'
                }
            });
        }
    }
}

export const clearCreateUser = () => dispatch => dispatch({type: actions.createUser.reset});

export const loginUser = async payload => {
    return async dispatch => {
        dispatch({
            type: actions.loginUser.request
        });

        try {
            const { data } = await axios.post('/api/login', payload);
            const { isAuth, error } = data;

            if (error) {
                dispatch({
                    type: actions.loginUser.receive,
                    payload: {
                        isAuth: false,
                        authError : error
                    }
                });

                return;
            }

            if (isAuth) {
                dispatch({
                    type: actions.loginUser.receive,
                    payload: {
                        isAuth,
                        error: null
                    }
                });
            }
        } catch (error) {
            dispatch({
                type: actions.loginUser.receive,
                payload: {
                    isAuth: false,
                    error
                }
            })
        }
    }
}

export function logOut() {
    const request = axios.get('/api/logout');

    return dispatch => {
        request.then( (response) => {
            if (response.status === 200) {
                cookie.remove('auth');
                
                dispatch({
                    type: actions.clearUserData
                });
                
                clearLogs();
                clearLastLogs();
            }
        })
    }
}

export function auth() {
    const response = axios.get('/api/auth')
    .then( ({data}) => data);

    return {
        type: actions.userAuth,
        payload: response
    }
}

export const editUserDetails = async ({userId, details}) => {
    const endpoint = '/api/editAccount';
    const body = {
        id : userId,
        ...details
    }

    return async dispatch => {
        dispatch({type: actions.editUser.request});

        try {
            const { data } = await axios.post(endpoint, body);
            const { error } = data;

            if (error) {
                dispatch({
                    type: actions.editUser.receive,
                    payload: {
                        isEditing: false,
                        success: false,
                        error: error.toString()
                    }
                });

                return
            }

            dispatch({
                type: actions.editUser.receive,
                payload: {
                    isEditing: false,
                    success: true,
                    error: null
                }
            });
        } catch (error) {
            dispatch({
                type: actions.editUser.receive,
                payload: {
                    isEditing: false,
                    success: false,
                    error: error.toString()
                }
            });
        }
    }
}

export const clearEditUser = () => dispatch => {dispatch({type: actions.editUser.reset})}

export const changePassword = async ({
    userId,
    currentPassword,
    newPassword
}) => {
    return async dispatch => {
        if (!userId || !currentPassword || !newPassword) {
            return;
        }

        const endpoint = '/api/changePassword';
        const body = {
            id : userId,
            oldPassword : currentPassword,
            newPassword
        }

        await dispatch({type: actions.changePassword.request});

        try {
            const { error } = await axios.post(endpoint, body);

            if (error) {
                dispatch({
                    type: actions.changePassword.receive,
                    payload: {
                        isChanging : false,
                        error : error,
                        success : false
                    }
                });

                return;
            }

            dispatch({
                type: actions.changePassword.receive,
                payload: {
                    isChanging : false,
                    error : null,
                    success : true
                }
            });
        } catch (error) {
            dispatch({
                type: actions.changePassword.receive,
                payload: {
                    isChanging : false,
                    error : error,
                    success : false
                }
            });
        }
    }
}

export const clearChangePassword = async () => async dispatch => {dispatch({type: actions.changePassword.reset})}

export const clearUserData = async () => async dispatch => {dispatch({type: actions.clearUserData})}