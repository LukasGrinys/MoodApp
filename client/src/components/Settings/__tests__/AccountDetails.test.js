import React from 'react';
import AccountDetails from '../AccountDetails';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { formErrorMessages } from '../../../constants/formErrorMessages';
import { editUserDetails, clearEditUser } from '../../../actions/user/asyncActions';

const mockStore = configureStore();
const testUserId = 'testID';
const accountDetailsProps = {
    email: 'test@mail.com',
    firstName: 'Test',
    lastName: 'Testington'
}

const store = mockStore({
    user: {
        id : testUserId
    }
});

const wrapper = mount(
    <Provider store={store}>
        <AccountDetails {...accountDetailsProps}/>
    </Provider>
);

describe('AccountDetails: initial UI', () => {
    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Shows users email', () => {
        const emailBlock = wrapper.find('.formWrapper').find('div').at(1);
        expect(emailBlock.text()).toBe(`User:  ${accountDetailsProps.email}`);
    });

    test('Has input field for editing firstName', () => {
        expect(wrapper.find('input[name="firstName"]').length).toBe(1);
    });

    test('firstName field has proper initial props', () => {
        expect(wrapper.find('input[name="firstName"]').props()).toEqual({
            className: "textInput",
            name: "firstName",
            onBlur: expect.any(Function),
            onChange: expect.any(Function),
            type: "text",
            value: accountDetailsProps.firstName
        });
    });

    test('Has input field for editing lastName', () => {
        expect(wrapper.find('input[name="lastName"]').length).toBe(1);
    });

    test('lastName field has proper initial props', () => {
        expect(wrapper.find('input[name="lastName"]').props()).toEqual({
            className: "textInput",
            name: "lastName",
            onBlur: expect.any(Function),
            onChange: expect.any(Function),
            type: "text",
            value: accountDetailsProps.lastName
        });
    });

    test('Has button for saving changes', () => {
        expect(wrapper.find('button').length).toBe(1);
    });

    test('Button has proper props', () => {
        expect(wrapper.find('button').props()).toEqual({
            className: 'button btnPrimary',
            onClick: expect.any(Function),
            type: 'button',
            children: 'Save changes'
        });
    });
});

describe('AccountDetails form logic', () => {
    test('Changes input value when typed', () => {
        const changedName = 'Teston';

        wrapper.find('input[name="firstName"]').simulate('change', {
            target : {
                name: 'firstName',
                value: changedName
            }
        });
        wrapper.update();

        expect(wrapper.find('input[name="firstName"]').prop('value')).toBe(changedName);
    });

    test('Shows error message if input is invalid', () => {
        wrapper.find('input[name="firstName"]').simulate('change', {
            target : {
                name: 'firstName',
                value: ''
            }
        });
        wrapper.update();

        setTimeout( () => {
            expect(wrapper.find('ErrorMessage').length).toBe(1);
            expect(wrapper.find('ErrorMessage').text()).toBe(formErrorMessages.required);
            done();
        }, 0);
    });

    test('Dispatches editUser action when clicked on button', () => {
        const firstName = wrapper.find('input[name="firstName"]').prop('value');
        const lastName = wrapper.find('input[name="lastName"]').prop('value');

        wrapper.find('button').simulate('click');
        wrapper.update();

        setTimeout( () => {
            expect(editUserDetails).toHaveBeenCalledTimes(1);
            expect(editUserDetails).toHaveBeenCalledWith({
                userId : testUserId,
                details : {
                    firstName,
                    lastName
                }
            });
        }, 0);
    })
});

describe('Handling success and errors', () => {
    test('Shows success box if the request was successful', () => {
        const store = mockStore({
            user: {
                editUser: {
                    success: true,
                    error: false
                }
            }
        });
        
        const successWrapper = mount(
            <Provider store={store}>
                <AccountDetails {...accountDetailsProps}/>
            </Provider>
        );

        expect(successWrapper.find('SuccessBox').length).toBe(1);
    });

    test('Shows error box if the request was not successful', () => {
        const testError = 'Test error';

        const store = mockStore({
            user: {
                editUser: {
                    success: false,
                    error: testError
                }
            }
        });
        
        const errorWrapper = mount(
            <Provider store={store}>
                <AccountDetails {...accountDetailsProps}/>
            </Provider>
        );

        expect(errorWrapper.find('ErrorBox').length).toBe(1);
        expect(errorWrapper.find('ErrorBox').text()).toBe(testError);
    });
})