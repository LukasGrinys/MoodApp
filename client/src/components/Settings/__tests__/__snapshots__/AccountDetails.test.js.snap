// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDetails: initial UI Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AccountDetails
    email="test@mail.com"
    firstName="Test"
    lastName="Testington"
  >
    <Wrapper>
      <div
        className="formWrapper"
      >
        <div>
          <strong>
            User: 
          </strong>
           
          test@mail.com
        </div>
        <br />
        <h3
          className="miniHeader"
        >
          Edit details:
        </h3>
        <Label>
          <span
            className="label"
          >
            First name:
          </span>
        </Label>
        <Input
          controlEvents={
            Object {
              "handleBlur": [Function],
              "handleChange": [Function],
            }
          }
          name="firstName"
          placeholder=""
          type="text"
          value="Test"
        >
          <input
            className="textInput"
            name="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="Test"
          />
        </Input>
        <Label>
          <span
            className="label"
          >
            Last name:
          </span>
        </Label>
        <Input
          controlEvents={
            Object {
              "handleBlur": [Function],
              "handleChange": [Function],
            }
          }
          name="lastName"
          placeholder=""
          type="text"
          value="Testington"
        >
          <input
            className="textInput"
            name="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="Testington"
          />
        </Input>
        <Footer>
          <div
            className="formFooter"
          >
            <Button
              handleClick={[Function]}
            >
              <button
                className="button btnPrimary"
                onClick={[Function]}
                type="button"
              >
                Save changes
              </button>
            </Button>
          </div>
        </Footer>
      </div>
    </Wrapper>
  </AccountDetails>
</Provider>
`;
